feat: pages_enabled admin controls with subscription-only search

NorkzYT opened the pull request:
## Summary

Adds `pages_enabled` admin controls for Popular, Trending, and Search. When YouTube search is disabled, users can still search within their subscriptions and saved playlists. Also prevents duplicate videos in playlists.

## Motivation

Invidious admins who run family instances need control over what users can access. A parent subscribes to approved channels on a child's account, disables search/trending/popular, and the child can only browse and search within those channels. No exposure to the wider YouTube catalog.

## Config
```yaml
pages_enabled:
  popular: false
  trending: false
  search: false
```

## Changes

### Admin Preferences Fix
- Checkboxes for Popular, Trending, and Search now persist when unchecked. The fallback defaulted to `"on"` because HTML checkboxes send no value when unchecked. Fixed to `"off"` to match the existing pattern for `captcha_enabled` and `login_enabled`.

### Route Blocking
- `/feed/popular`, `/feed/trending`, `/hashtag/*` return 403 when disabled.

- `/api/v1/search`, `/api/v1/search/suggestions`, `/api/v1/hashtag/*` return 403 JSON when disabled.

### Subscription-Only Search
- When `pages_enabled.search` is `false` and the user is logged in, `/search` queries the subscription materialized view and `playlist_videos` table instead of YouTube.
- Results merge, deduplicate by video ID, and sort by publish date.
- Anonymous users see a 403 with a login prompt.
- Search bar placeholder changes to "Search subscriptions & playlists..."
- YouTube search filters (date, type, duration) are hidden.
- Results page shows a lock icon with "Searching within your subscriptions and playlists only."

### Playlist Duplicate Prevention
- Adding a video that already exists in a playlist is now a no-op on the web UI and returns 409 via the API.
- Applied to the web route (`/playlist_ajax`) and the API route (`POST /api/v1/auth/playlists/:plid/videos`).

### UI

- Search bar is always visible (subscription search needs it).
- Search homepage shows a hint when YouTube search is disabled.
- Feed menu hides disabled pages.

## Files Changed

- `src/invidious/routes/preferences.cr` — checkbox persistence fix
- `src/invidious/routes/before_all.cr` — route blocking logic
- `src/invidious/routes/search.cr` — subscription-only search mode
- `src/invidious/search/processors.cr` — `subscriptions_and_playlists` search method
- `src/invidious/routes/playlists.cr` — playlist duplicate check (web)
- `src/invidious/routes/api/v1/authenticated.cr` — playlist duplicate check (API)
- `src/invidious/views/template.ecr` — navbar search bar
- `src/invidious/views/search_homepage.ecr` — subscription hint
- `src/invidious/views/search.ecr` — conditional filters and empty state
- `src/invidious/views/components/search_box.ecr` — dynamic placeholder
- `locales/en-US.json` — 6 new locale strings

## Screenshots

<img width="1920" height="935" alt="image" src="https://github.com/user-attachments/assets/b4fef042-d511-4cef-bb2a-e31c50c65009" />

<img width="1920" height="935" alt="image" src="https://github.com/user-attachments/assets/cf122b0b-45a6-424c-9c6b-e0c3509b975f" />

<img width="1920" height="935" alt="image" src="https://github.com/user-attachments/assets/1965e52d-3fa4-4f32-a2b2-2494f3b295e4" />

<img width="1920" height="935" alt="image" src="https://github.com/user-attachments/assets/d02720ae-f6cb-46c4-bf39-6914d0dc28c6" />

<img width="1920" height="929" alt="image" src="https://github.com/user-attachments/assets/fe6fb4bc-6d34-4cd8-9803-6f4c06582492" />

<img width="1920" height="935" alt="image" src="https://github.com/user-attachments/assets/223ec4ea-311f-4932-93fd-02410ca6b72e" />

<img width="1920" height="935" alt="image" src="https://github.com/user-attachments/assets/1240db50-3675-42b9-9a82-20987e8751a9" />

<img width="1920" height="935" alt="image" src="https://github.com/user-attachments/assets/22fd16e6-cbfd-4463-823e-58e5598d36ee" />

Commit:
feat: optional `trending_enabled` and `search_enabled` params added

iBicha commented on src/invidious/routes/api/v1/feeds.cr:
```diff
@@ -4,6 +4,11 @@ module Invidious::Routes::API::V1::Feeds
 
+    if !CONFIG.trending_enabled
+      error_message = {"error" => "Administrator has disabled this endpoint."}.to_json
+      haltf env, 400, error_message
```
FYI https://github.com/iv-org/invidious/pull/4296/files

NorkzYT replied:
My mistake, fixing now.

NorkzYT committed:
fix(feeds.cr): http status code

Commit:
fix(search.cr): http status code

NorkzYT commented:
Code formatting issues will be fixed soon.

Commit:
style(feeds.cr): fix code formatting

unixfox commented:
If we are going to allow every specific pages to be disabled, we might as well use an array for that like:

```
pages_enabled:
  trending: true
  popular: true
  search: true
```

NorkzYT commented:
@unixfox 

Great suggestion. Do you recommend I implement it in this GitHub Pull Request or in a separate one?

unixfox commented:
Would prefer in this one but we would need to keep compatibility with existing popular_enabled parameter.

NorkzYT commented:
Understood, that works.

unixfox commented:
@SamantazFox advice on my comment please: https://github.com/iv-org/invidious/pull/4674#issuecomment-2118875288

SamantazFox commented:
> If we are going to allow every specific pages to be disabled, we might as well use an array for that like:
> 
> ```
> pages_enabled:
>   trending: true
>   popular: true
>   search: true
> ```

Yes, I second that! However, it will be a breaking change (will go with the next major version release).

@NorkzYT Though, why do you want to disable search? For the trending feed, I can understand that you may not want random crap in your home page, but why search?

NorkzYT commented:
@SamantazFox 

My plan is to enable Invidious admins to control which features are accessible to users. This is particularly important for families with young children who prefer YouTube over YouTube Kids. I found that existing tools lacked the ability to give admins such granular control over user features, thus I decided to contribute to Invidious by developing these admin control features.

For instance, in my case, I want to restrict access for some users that only very young children use to only allow certain YouTube channels that are pre-approved and subscribed to their user account. By disabling features like search, trending, and popular, I can ensure that children only interact with videos from these subscribed channels and playlists, thereby preventing exposure to unsuitable and overage content.

Additionally, I propose adding a search bar specifically for subscriptions and playlists. This would allow users to quickly find videos within these predefined categories without the need to navigate through entire channels or playlists, thus saving time and ensuring a safer viewing experience.

I know that a tool called TubeArchivist has a search feature for videos that have already been downloaded. However, I'm switching from [TubeArchivist](https://github.com/tubearchivist/tubearchivist) to Invidious because I no longer want to host 5TB of downloaded YouTube videos, especially as the library keeps growing.

Commit:
feat(invidious): specific pages are disabled with the use of an array

Commit:
chore(search.cr): add newline at end of file

NorkzYT commented:
Complete. Someone please test as the only issue I know is the Invidious docker container becoming `unhealthy` if I specify `pages_enabled` in the environmental variables under `INVIDIOUS_CONFIG:` as shown below even though it works without issues in Invidious itself.

```docker
services:
  invidious:
    container_name: invidious
    hostname: invidious
    build:
      context: .
      dockerfile: docker/Dockerfile.arm64
    restart: unless-stopped
    ports:
      - "3000:3000"
    environment:
      # Please read the following file for a comprehensive list of all available
      # configuration options and their associated syntax:
      # https://github.com/iv-org/invidious/blob/master/config/config.example.yml
      INVIDIOUS_CONFIG: |
        db:
          dbname: invidious
          user: kemal
          password: kemal
          host: invidious-db
          port: 5432
        check_tables: true
        # external_port:
        # domain:
        # https_only: false
        # statistics_enabled: false
        hmac_key: "xxx"
        player_style: youtube
        save_player_pos: true
        registration_enabled: true
        pages_enabled:
          popular: false
          trending: false
          search: false
        captcha_enabled: false
        feed_menu: ["Subscriptions", "Playlists"]
        default_home: <none>
        related_videos: false
        admins: ["norkz"] # set your user name
    healthcheck:
      test: wget -nv --tries=1 --spider http://127.0.0.1:3000/api/v1/trending || exit 1
      interval: 30s
      timeout: 5s
      retries: 2
    networks:
      - proxy
```

iBicha commented:
> Complete. Someone please test as the only issue I know is the Invidious docker container becoming `unhealthy` if I specify `pages_enabled` in the environmental variables under `INVIDIOUS_CONFIG:` as shown below even though it works without issues in Invidious itself.

The health check is hitting the trending endpoint
```
    healthcheck:
      test: wget -nv --tries=1 --spider http://127.0.0.1:3000/api/v1/trending || exit 1
```
Which you have disabled, so it is natural for it to appear unhealthy

NorkzYT commented:
@iBicha 

Ah, thank you for the explanation. Much appreciated. I went right by it and did not notice.

Commit:
chore(lint): format with crystal tool

SamantazFox commented on config/config.example.yml:
```diff
@@ -203,12 +203,13 @@ https_only: false
 
 ##
-## Enable/Disable the "Popular" tab on the main page.
+## Enable/Disable specific pages on the main page.
+## Example:
+## pages_enabled:
+##   trending: true
+##   popular: true
+##   search: true
```
To remain consistent with the rest of the file (a single `#` is used to show the default value of a config option that is not mandatory):
```suggestion
## Enable/Disable specific pages on the main page.
##
#pages_enabled:
#  trending: true
#  popular: true
#  search: true
```

SamantazFox commented on src/invidious/views/user/preferences.ecr:
```diff
@@ -284,9 +284,18 @@
+                    <input name="popular_enabled" id="popular_enabled" type="checkbox" <% if CONFIG.page_enabled?("popular") %>checked<% end %>>
                 </div>
 
+                <div class="pure-control-group">
+                    <label for="trending_enabled"><%= translate(locale, "Trending enabled: ") %></label>
+                    <input name="trending_enabled" id="trending_enabled" type="checkbox" <% if CONFIG.page_enabled?("trending") %>checked<% end %>>
+                </div>
+
+                <div class="pure-control-group">
+                    <label for="search_enabled"><%= translate(locale, "Search enabled: ") %></label>
+                    <input name="search_enabled" id="search_enabled" type="checkbox" <% if CONFIG.page_enabled?("search") %>checked<% end %>>
+                </div>
```
We are trying to unify the naming/format of translation keys (the string that is being searched to get the associated translated string):

```suggestion
                <div class="pure-control-group">
                    <label for="trending_enabled"><%= translate(locale, "preferences_trending_enabled_label") %></label>
                    <input name="trending_enabled" id="trending_enabled" type="checkbox" <% if CONFIG.page_enabled?("trending") %>checked<% end %>>
                </div>

                <div class="pure-control-group">
                    <label for="search_enabled"><%= translate(locale, "preferences_search_enabled_label") %></label>
                    <input name="search_enabled" id="search_enabled" type="checkbox" <% if CONFIG.page_enabled?("search") %>checked<% end %>>
                </div>
```

In all cases, you need to add these key/value combinations to the end of `locales/en_US.json` for them to be translated in other languages:

```json
{
    ...
    "some_other_key": "Some other string",
    "preferences_trending_enabled_label": "Trending enabled: ",
    "preferences_search_enabled_label": "Search enabled: "
}
```

NorkzYT committed:
Update config/config.example.yml

Co-authored-by: Samantaz Fox <coding@samantaz.fr>

Commit:
feat(en-US.json): add translation key and value

NorkzYT committed:
Update src/invidious/views/user/preferences.ecr

Co-authored-by: Samantaz Fox <coding@samantaz.fr>

NorkzYT commented:
@SamantazFox 

Thank you for the info. Suggestions have been added and completed.

NorkzYT committed:
Merge branch 'master' into optional-disable-api-features

NorkzYT committed:
Merge branch 'iv-org:master' into optional-disable-api-features

NorkzYT commented:
Any update on merging this?
@SamantazFox

deggle commented:
Any news on this? Disabling the search is a bit of a deal breaker for me being able to let my kids access YT.

syeopite commented on src/invidious/config.cr:
```diff
@@ -100,7 +100,7 @@ class Config
   # Subscribe to channels using PubSubHubbub (requires domain, hmac_key)
   property use_pubsub_feeds : Bool | Int32 = false
-  property popular_enabled : Bool = true
+  property pages_enabled : Hash(String, Bool) = {"trending" => true, "popular" => true, "search" => true}
```
I think the primary blocker for this PR right now is that this is a breaking change. I'll suggest to keep `popular_enabled` for now but maybe with a deprecation warning. 

Some clear precedence of the two config options should also be set and documented to handle the case of someone configuring both of them at the same time.

syeopite reviewed (changes requested):
There is a lot of lines changed across the route handlers for whats effectively a pretty tiny change. 

I think it'll make more sense to apply this either on the route definitions themselves in `src/invidious/routing.cr` or just on the before routes handler in `src/invidious/routes/before_all.cr`

NorkzYT committed:
Merge branch 'master' into optional-disable-api-features

Commit:
feat: test

Commit:
feat: test2

NorkzYT commented:
I have made some changes for testing purposes and am now awaiting for https://github.com/iv-org/invidious/pull/5312 to be merged since on the latest commit I am receiving the following:

It looks like you found a bug in Invidious!


Before reporting a bug, make sure that you have:

tried to [refresh the page](http://100.97.39.119:3000/watch?v=s6nXIpOLQz0)
tried to [use another instance](https://redirect.invidious.io/watch?v=s6nXIpOLQz0)
read the [Frequently Asked Questions (FAQ)](https://github.com/iv-org/documentation/blob/master/docs/faq.md)
searched for [existing issues on GitHub](https://github.com/iv-org/invidious/issues?q=is:issue+is:open+%5BBug%5D+Missing+hash+key%3A+%22videoDetails%22+%28KeyError%29)

If none of the above helped, please [open a new issue on GitHub](https://github.com/iv-org/invidious/issues/new?labels=bug&template=bug_report.md&title=%5BBug%5D+Missing+hash+key%3A+%22videoDetails%22+%28KeyError%29) (preferably in English) and include the following text in your message (do NOT translate that text):

Title: `Missing hash key: "videoDetails" (KeyError)`
Date: `2025-05-22T12:44:52Z`
Route: `/watch?v=s6nXIpOLQz0`
Version: `2025.05.17-f07859e @ (HEAD detached at v2.20250517.0)`

<details>
<summary>Backtrace</summary>
<p>
   
```
Missing hash key: "videoDetails" (KeyError)
  from /usr/share/crystal/src/hash.cr:1151:9 in 'dig'
  from /usr/share/crystal/src/json/any.cr:342:3 in 'extract_video_info'
  from src/invidious/videos.cr:327:10 in 'fetch_video'
  from src/invidious/videos.cr:315:13 in 'get_video:region'
  from src/invidious/routes/watch.cr:55:15 in 'handle'
  from lib/kemal/src/kemal/route.cr:13:9 in '->'
  from src/invidious/helpers/handlers.cr:31:37 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:30:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:30:7 in 'call_next'
  from lib/kemal/src/kemal/filter_handler.cr:21:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:30:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:30:7 in 'call_next'
  from src/invidious/helpers/handlers.cr:95:12 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:30:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:30:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:30:7 in 'call_next'
  from src/ext/kemal_static_file_handler.cr:106:14 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:30:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:30:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:30:7 in 'call'
  from /usr/share/crystal/src/http/server/request_processor.cr:51:11 in 'handle_client'
  from /usr/share/crystal/src/fiber.cr:143:11 in 'run'
  from ???
```
</p>
</details>

Fijxu commented:
>I have made some changes for testing purposes and am now awaiting for https://github.com/iv-org/invidious/pull/5312 to be merged since on the latest commit I am receiving the following:

Maybe setup `inv_sig_helper` or invidious companion in your Invidious development instance so you can keep working on this without having to wait for the PR to be merged

NorkzYT commented:
@Fijxu 

I have and the same issue persists as someone else advised in https://github.com/iv-org/invidious/issues/5307.

Here is my docker compose file:


<details>
<summary>docker-compose.yml</summary>

```yml
# Warning: This docker-compose file is made for development purposes.
# Using it will build an image from the locally cloned repository.
#
# If you want to use Invidious in production, see the docker-compose.yml file provided
# in the installation documentation: https://docs.invidious.io/installation/

services:
  invidious:
    container_name: invidious
    hostname: invidious
    image: quay.io/invidious/invidious:latest
    restart: unless-stopped
    ports:
      - "3000:3000"
    environment:
      # Please read the following file for a comprehensive list of all available
      # configuration options and their associated syntax:
      # https://github.com/iv-org/invidious/blob/master/config/config.example.yml
      INVIDIOUS_CONFIG: |
        db:
          dbname: invidious
          user: kemal
          password: kemal
          host: invidious-db
          port: 5432
        check_tables: true
        signature_server: inv_sig_helper:12999
        visitor_data: "xxx"
        po_token: "xxx"
        hmac_key: "xxx"
        player_style: youtube
        save_player_pos: true
        registration_enabled: true
        related_videos: false
        admins: ["norkz"] # set your user name
    healthcheck:
      test: wget -nv --tries=1 --spider http://127.0.0.1:3000/api/v1/trending || exit 1
      interval: 30s
      timeout: 5s
      retries: 2
    logging:
      options:
        max-size: "1G"
        max-file: "4"
    depends_on:
      - invidious-db
    networks:
      - proxy

  inv_sig_helper:
    image: quay.io/invidious/inv-sig-helper:latest
    init: true
    command: ["--tcp", "0.0.0.0:12999"]
    environment:
      - RUST_LOG=info
    restart: unless-stopped
    cap_drop:
      - ALL
    read_only: true
    security_opt:
      - no-new-privileges:true
    networks:
      - proxy

  invidious-db:
    container_name: invidious-db
    hostname: invidious-db
    image: docker.io/library/postgres:14
    restart: unless-stopped
    volumes:
      - postgresdata:/var/lib/postgresql/data
      - ./config/sql:/config/sql
      - ./docker/init-invidious-db.sh:/docker-entrypoint-initdb.d/init-invidious-db.sh
    environment:
      POSTGRES_DB: invidious
      POSTGRES_USER: kemal
      POSTGRES_PASSWORD: kemal
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U $$POSTGRES_USER -d $$POSTGRES_DB"]
      interval: 5s
      timeout: 5s
      retries: 5
    networks:
      - proxy

volumes:
  postgresdata: {}

networks:
  proxy:
    driver: bridge
    external: true
```

</details>

NorkzYT commented:
Understood.

NorkzYT committed:
Update locales/en-US.json

Co-authored-by: Fijxu <fijxu@nadeko.net>

NorkzYT committed:
Update locales/en-US.json

Co-authored-by: Fijxu <fijxu@nadeko.net>

NorkzYT committed:
📝 (locales/en-US.json): remove redundant translation strings

NorkzYT committed:
🌐 (feeds.cr, search.cr): update translation keys for disabled feeds and search messages

NorkzYT commented:
Done. I had forgotten about that.

NorkzYT commented:
What else needs to be done for this PR to be completed and ready for review?

syeopite commented on src/invidious/routes/preferences.cr:
```diff
@@ -199,9 +199,12 @@ module Invidious::Routes::PreferencesRoute
-        popular_enabled = env.params.body["popular_enabled"]?.try &.as(String)
-        popular_enabled ||= "off"
-        CONFIG.popular_enabled = popular_enabled == "on"
+        pages_enabled = {
+          "popular"  => (env.params.body["popular_enabled"]?.try &.as(String) || "off") == "on",
+          "trending" => (env.params.body["trending_enabled"]?.try &.as(String) || "off") == "on",
+          "search"   => (env.params.body["search_enabled"]?.try &.as(String) || "off") == "on",
+        }
+        CONFIG.pages_enabled = pages_enabled
```
Shouldn't the fallback value be `on` instead?

```suggestion
        pages_enabled = {
          "popular"  => (env.params.body["popular_enabled"]?.try &.as(String) || "on") == "on",
          "trending" => (env.params.body["trending_enabled"]?.try &.as(String) || "on") == "on",
          "search"   => (env.params.body["search_enabled"]?.try &.as(String) || "on") == "on",
        }
        CONFIG.pages_enabled = pages_enabled
```

syeopite commented on src/invidious/config.cr:
```diff
@@ -111,13 +111,29 @@ class Config
+  # Global per-page feature toggles.
+  # Valid keys: "trending", "popular", "search"
+  # If someone sets both `popular_enabled` and `pages_enabled["popular"]`, the latter takes precedence.
+  property pages_enabled : Hash(String, Bool) = {
+    "trending" => true,
+    "popular"  => true,
+    "search"   => true,
+  }
```
Please make this into a separate struct rather than a hash

syeopite commented:
Just did a second review. Please also the move route changes to the before all routes handler over at  `src/invidious/routes/before_all.cr`. The checks are pretty minor and generalized so I don't believe it needs to be handled at each of the routes themselves.

syeopite commented on src/invidious.cr:
```diff
@@ -191,7 +191,7 @@ if (CONFIG.use_pubsub_feeds.is_a?(Bool) && CONFIG.use_pubsub_feeds.as(Bool)) ||
 end
 
-if CONFIG.popular_enabled
+if CONFIG.page_enabled?("popular")
   Invidious::Jobs.register Invidious::Jobs::PullPopularVideosJob.new(PG_DB)
 end
```
(**Not a blocking review just a comment**): Huh it looks like the popular page will stay empty if an admin enables the popular feed through the preferences page rather than the config at start-up, since the job will never be started.

This isn't something that's caused by the PR but rather Invidious' current behavior. 

It could be fixed via a select statement and some global channels but it should probably be handled by some sort of global job scheduler instead 🤔

NorkzYT committed:
feat(config.cr): introduce PagesEnabled struct for managing feature toggles for pages
refactor(routes): replace direct page_enabled checks with centralized logic in before_all.cr for cleaner endpoint management
chore(routes): remove redundant page_enabled checks from individual routes to streamline code and improve maintainability

NorkzYT committed:
Merge branch 'iv-org:master' into optional-disable-api-features

syeopite committed:
Config: Use from_yaml constructor for PagesEnabled

`PagesEnabled` doesn't define a blank constructor without any
parameters meaning that `PagesEnabled.new` leads to a error due to lack
of arguments

syeopite committed:
Use `PagesEnabled` struct when setting pages_enabled

The config update logic for the admin panel was not updated to use
the `PagesEnabled` struct introduced in commit `e238624`

syeopite committed:
Fix config precedence with popular_enabled

The original approach cannot be used to properly handle the precedence
of `popular_enabled` in relations to the new `pages_enabled` since it
is impossible to determine whether they are unset due to the presence
of default values.

In addition the original precedence handling wasn't actually doing
anything since PagesEnabled.has_key? will always return true if the key
is valid and thus use page_enabled and its default values regardless of
whether popular_enabled was set or not (which again is also impossible
to know)

To fix this two new instance variables are introduced in order to track
whether these two attributes were present in the YAML config and the
logic for handling them is instead reduced to only using `page_enabled`
and copying `popular_enabled` to page_enabled["popular"] when
`page_enabled` field is unset.

syeopite committed:
Mark attributes set over env var as present if needed

syeopite committed:
Add tests for `popular_enabled` deprecation logic

NorkzYT committed:
chore: disable trending by default

NorkzYT committed:
chore: missing hash key issue

NorkzYT committed:
latest from master

NorkzYT committed:
Merge branch 'master' into optional-disable-api-features

NorkzYT committed:
revert docker compose file

NorkzYT commented:
The following config is not working at the moment. Will continue working on this.

```
pages_enabled:
  trending: true
  popular: true
  search: true
```

NorkzYT committed:
Merge branch 'master' into optional-disable-api-features

NorkzYT committed:
feat: add search_page_disabled locale + preferences page filtering + debug config route

- Fix kemal_static_file_handler directory_listing to use Path types (Crystal API compat)
- Add missing search_page_disabled translation key to en-US.json
- Filter preferences default_home options based on pages_enabled config
- Add admin-only /debug_config route for verifying configuration state

NorkzYT committed:
fix(preferences): unchecked pages_enabled checkboxes now persist correctly

HTML checkboxes send no value when unchecked, so the fallback
"|| \"on\"" always re-enabled popular/trending/search on save.
Changed default to "off" to match the pattern used by captcha_enabled,
login_enabled, and other checkbox toggles in the same handler.

NorkzYT committed:
fix(pages_enabled): hide search UI and block all search routes when disabled

- Hide navbar search box when CONFIG.page_enabled?("search") is false
- Keep Invidious logo visible with proper width when search bar is hidden
- Hide search widget on search_homepage when search is disabled
- Block /results, /api/v1/search/suggestions in before_all guard
- Block /hashtag/* and /api/v1/hashtag/* when search is disabled

NorkzYT committed:
feat(search): subscription-only search mode when YouTube search is disabled

When pages_enabled.search is false, instead of blocking search entirely,
the search bar now searches within the user's subscriptions and saved
playlists only. YouTube's search API is never called.

Changes:
- routes/search.cr: detect search_disabled, force subscription+playlist
  search for logged-in users, 403 for anonymous users
- search/processors.cr: new subscriptions_and_playlists() method that
  queries both the subscription materialized view and playlist_videos
  table, merges and deduplicates results
- before_all.cr: let /search through to the route handler (APIs still
  blocked), keep /hashtag/* blocked
- search_box.ecr: dynamic placeholder ("Search subscriptions & playlists")
- search_homepage.ecr: hint text when in subscription-only mode
- search.ecr: hide YouTube filters, show lock icon + notice, custom
  empty-results message
- template.ecr: always show search bar (subscription search needs it)
- en-US.json: 6 new locale strings for subscription-only search UX

NorkzYT committed:
fix(playlists): prevent duplicate videos in the same playlist

Check if the video already exists in the playlist before inserting.
Uses the existing select_index() query. Returns 409 via the API and
silently redirects back on the web UI.

Applied to both the web route (playlist_ajax) and the API v1 route
(POST /api/v1/auth/playlists/:plid/videos).

NorkzYT committed:
fix(search_homepage): center subscription hint text under search bar

NorkzYT committed:
fix(before_all): halt response after rendering disabled page error

The return from before_all did not stop the route handler from
executing. The trending/popular/hashtag routes still ran and
overwrote the 403 response. Now set a halted flag and use Kemal
halt in the before_all block to prevent route handler execution.

Error page now renders with the full Invidious theme (dark mode,
navbar, CSS) matching the rest of the UI.

NorkzYT commented:
GitHub PR Complete.

Title and Description updated to explain the changes in this PR.

NorkzYT committed:
Merge branch 'master' into optional-disable-api-features

NorkzYT committed:
fix(before_all): use I18n.translate for page_disabled error messages

NorkzYT committed:
fix(search_box): use I18n.translate instead of bare translate in ECR template

NorkzYT committed:
fix: replace bare translate() with I18n.translate() in ECR templates and routes

NorkzYT committed:
fix(before_all): avoid double-write on halted responses to prevent HTTP/2 corruption

Store response body in env instead of writing directly; let halt() do the single
write+close in the before_all block, eliminating Closed stream errors and
ERR_HTTP2_PROTOCOL_ERROR on subsequent requests (e.g. /login).

NorkzYT committed:
fix(extractors): skip lockupViewModel items missing collectionThumbnailViewModel

YouTube changed the thumbnail structure for some collection/playlist items.
Use dig? and return nil to skip unparseable items instead of crashing.

NorkzYT committed:
Merge branch 'master' into optional-disable-api-features

NorkzYT committed:
fix(extractors): restore collectionThumbnailViewModel path with dig? fallback

Try collectionThumbnailViewModel path first, fall back to direct
thumbnailViewModel for lockupViewModel items that use different structures.
Items with neither path are skipped cleanly.

NorkzYT committed:
fix(playlists): fall back to playlistHeaderRenderer when sidebar is absent

YouTube stopped returning sidebar.playlistSidebarRenderer for many
playlists; extract the same fields from header.playlistHeaderRenderer
instead to prevent 500 errors on /playlist pages.

Also fix the -1 video count on lockupViewModel playlist cards: make
badge text matching case-insensitive and add a numeric-prefix fallback
so counts are recovered even when YouTube changes the badge wording.

NorkzYT committed:
fix(extractors): distinguish video vs playlist in LockupViewModelParser

Video IDs are always 11 characters; playlists have a playlistId in their
watchEndpoint. Previously all lockupViewModel items were returned as
SearchPlaylist, causing video links on channel pages to route to
/playlist?list= instead of /watch?v=.

Authored-by: NorkzYT <richard@pcscorp.dev>
